docs/proposals/20250513-propogate-taints.md

```diff
@@ -0,0 +1,455 @@
+
+To support taints for control plane Machines, the contract for ControlPlanes should be extended.
+
+Similar to the extension for `ReadinessGates` at the `Machines` section, an additional optional  `FooControlPlaneMachineTemplate` rule should be added for supporting taints.
+
+Support and usage of that field would result in having the defined taints set on the Machines created by the ControlPlane provider and the Machine controller being responsible to behave as for worker Machines.
+
+#### Proposed controller changes
+
+##### Changes to the Machine and MachinePool controllers
+
+The Machine and MachinePool controllers as of today already manage the lifecycle of the taint `node.cluster.x-k8s.io/uninitialized`.
+
+These controllers should make use of that taint and add all `Initialize` taints at the same time as they remove it from a corresponding Node.
+This ensures that a node ever only gets these taints added once at the time of finishing initialization of a Node.
+
+If a Node re-registers itself and this includes having the `node.cluster.x-k8s.io/uninitialized`, the controllers also would again add the `Initialize` taints.
+
+Also all `Always` taints should get set for the first time, before or at the time of the removal of `node.cluster.x-k8s.io/uninitialized`.
+
+Also at this stage the controllers could check and (re-)add all taints of the `Always` type.
+
+Note: With this change we should also add the information to the "Contract rules for BootstrapConfig" documentation, that `Initialize` taints are only added when a bootstrap provider supports  the `node.cluster.x-k8s.io/uninitialized`.
+
+To keep track of which taints have been added by the controller, the controller also adds the annotation `cluster.x-k8s.io/taints-from-machine` to the Node.
+This follows the convention established by `cluster.x-k8s.io/labels-from-machine`.
+
+The taints will be concatenated with `,` and the serialization will look like this given the 4 ways a taint can be specified:
+`cluster.x-k8s.io/taints-from-machine:<key1>=<value1>:<effect1>,<key2>=<value2>:,<key3>:<effect3>,<key4>`
+
+See the upstream [string implementation](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/api/core/v1/taint.go) for more details.
+
+##### Changes to the Cluster topology, MachineSet and MachineDeployment controllers
+
+The Cluster's topology controller will properly calculate and in-place propagate the new introduced taint fields down to the ControlPlane object (if it complies with the optional contract part), MachineDeployment and/or MachinePool resources.
+
+The MachineDeployment and MachineSet controllers will watch their associated `MachineSpec` for updates to the `spec.taints` field and update their owned objects objects (MachineSet or Machine) to reflect the new values.
+
+This should be done similar to how the existing in-place mutable fields like `ReadinessGates`, `NodeDrainTimeout`, etc. for a Machine are handled today.
+
```

Comment:
Can we add a short section to mention cluster-autoscaler changes? (I think there is something that needs to be done that autoscaler can use the taints, we don't have to go in details how that would be done in the autoscaler, just to highlight that there's work to be done)

I think you can just mention the `capacity.cluster-autoscaler.kubernetes.io/taints` annotation (https://cluster-api.sigs.k8s.io/tasks/automated-machine-management/autoscaling#pre-defined-labels-and-taints-on-nodes-scaled-from-zero) and say that autoscaler would directly consume the taints field if it is set? (let's defer details like overwrite mechanism between field and annotations)